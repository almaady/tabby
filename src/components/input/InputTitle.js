import React from "react";

import PropTypes from "prop-types";

import autosize from "autosize";

import classNames from "classnames";

/**
 * @class InputTitle - It is an input for titles where they look as text
 */
class InputTitle extends React.PureComponent {
    /**
   * @property {string} placeholder -The placeholder of the input
   * @property {string} className - The type of the input
   * @property {string} initialText- The placeholder of the input
   * @property {number} maxLength - The number of characters
   * @property {bool} readOnly - Determines if you can edit text
   */
  static propTypes = {
    initialText: PropTypes.string,
    className: PropTypes.string,
    placeholder: PropTypes.string,
    onTextEdited: PropTypes.func.isRequired,
    allowBlank: PropTypes.bool.isRequired,
    resetOnFinish: PropTypes.bool.isRequired,
    maxLength: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    readOnly: PropTypes.bool.isRequired
  };
  constructor(props) {
    super(props);

    this.state = {
      text: props.initialText
    };

    this.textInput = null;
  }

  setTextInputRef = element => {
    this.textInput = element;
  };

  setTextTrim = () => {
    if (!this.props.allowBlank && this.textInput.value.trim() === "") {
      return;
    }

    this.setState({
      text: this.textInput.value.trim()
    });

    this.props.onTextEdited(this.textInput.value.trim());

    if (this.props.resetOnFinish) {
      this.textInput.value = "";
    }
  };

  onKeyDown = event => {
    if (event.keyCode === 13) {
      this.textInput.blur();
      event.preventDefault();
    }

    autosize(this.textInput);
  };

  componentDidMount() {
    autosize(this.textInput);
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.initialText !== this.props.initialText) {
      this.textInput.value = this.props.initialText;
      autosize(this.textInput)
    }
  }

  render() {
    const {
      placeholder,
      className,
      initialText,
      maxLength,
      readOnly
    } = this.props;
    return (
      <textarea
        placeholder={placeholder}
        onBlur={this.setTextTrim}
        className={classNames("textarea__edition", className)}
        ref={this.setTextInputRef}
        onKeyDown={this.onKeyDown}
        maxLength={maxLength}
        defaultValue={initialText}
        rows="5"
        readOnly={readOnly}
      />
    );
  }
}

InputTitle.defaultProps = {
  allowBlank: true,
  resetOnFinish: false,
  readOnly: false,
  maxLength: "",
  className: "",
  placeholder: "",
  initialText: ""
};

export default InputTitle;
