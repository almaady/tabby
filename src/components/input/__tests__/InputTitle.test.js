import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";

import InputTitle from "../InputTitle";

describe("Inputtitle", () => {
  const onTextEdited = sinon.stub();
  let props;
  beforeEach(() => {
    props = {
      resetOnFinish: true,
      readOnly: false,
      onTextEdited
    };
  });
  it("should render correctly with default state", () => {
    const component = shallow(<InputTitle {...props} />);
    expect(component).toMatchSnapshot();
  });
  it("should trim text from input value", () => {
    const component = mount(<InputTitle {...props} />);
    const instance = component.instance();
    instance.textInput.value = " without spaces ";
    component.simulate("blur");
    expect(component.state().text).toEqual("without spaces");
  });
  it("should return nothing", () => {
    const _onTextEdited = sinon.spy();
    const component = mount(
      <InputTitle {...props} allowBlank={false} onTextEdited={_onTextEdited} />
    );
    const instance = component.instance();
    instance.textInput.value = "";
    component.simulate("blur");
    expect(_onTextEdited.called).toEqual(false);
  });
  it("should return nothing", () => {
    const component = mount(<InputTitle {...props} resetOnFinish={true} />);
    const instance = component.instance();
    instance.textInput.value = "Value";
    component.simulate("blur");
    expect(instance.textInput.value).toEqual("");
  });
  it("should call onBlur text from input value", () => {
    const component = mount(<InputTitle {...props} />);
    const instance = component.instance();
    instance.textInput.blur = sinon.spy();
    component.simulate("keyDown", { keyCode: 13 });
    expect(instance.textInput.blur.called).toBeTruthy();
  });
});
