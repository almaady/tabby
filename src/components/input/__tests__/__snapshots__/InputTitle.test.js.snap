// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputtitle should render correctly with default state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InputTitle
    allowBlank={true}
    className=""
    initialText=""
    maxLength=""
    onTextEdited={[Function]}
    placeholder=""
    readOnly={false}
    resetOnFinish={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "textarea__edition",
      "defaultValue": "",
      "maxLength": "",
      "onBlur": [Function],
      "onKeyDown": [Function],
      "placeholder": "",
      "readOnly": false,
      "rows": "5",
    },
    "ref": [Function],
    "rendered": null,
    "type": "textarea",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "textarea__edition",
        "defaultValue": "",
        "maxLength": "",
        "onBlur": [Function],
        "onKeyDown": [Function],
        "placeholder": "",
        "readOnly": false,
        "rows": "5",
      },
      "ref": [Function],
      "rendered": null,
      "type": "textarea",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
